core/lib/src/local/blocking/request.rs

```diff
@@ -0,0 +1,384 @@
+/// A structure representing a local request as created by [`Client`].
+///
+/// # Usage
+///
+/// A `LocalRequest` value is constructed via method constructors on [`Client`].
+/// Headers can be added via the [`header`] builder method and the
+/// [`add_header`] method. Cookies can be added via the [`cookie`] builder
+/// method. The remote IP address can be set via the [`remote`] builder method.
+/// The body of the request can be set via the [`body`] builder method or
+/// [`set_body`] method.
+///
+/// ## Example
+///
+/// The following snippet uses the available builder methods to construct a
+/// `POST` request to `/` with a JSON body:
+///
+/// ```rust
+/// use rocket::local::blocking::Client;
+/// use rocket::http::{ContentType, Cookie};
+///
+/// let client = Client::new(rocket::ignite()).expect("valid rocket");
+/// let req = client.post("/")
+///     .header(ContentType::JSON)
+///     .remote("127.0.0.1:8000".parse().unwrap())
+///     .cookie(Cookie::new("name", "value"))
+///     .body(r#"{ "value": 42 }"#);
+/// ```
+///
+/// # Dispatching
+///
+/// A `LocalRequest` can be dispatched in one of two ways:
+///
+///   1. [`dispatch`]
+///
+///      This method should always be preferred. The `LocalRequest` is consumed
+///      and a response is returned.
+///
+///   2. [`mut_dispatch`]
+///
+///      This method should _only_ be used when either it is known that the
+///      application will not modify the request, or it is desired to see
+///      modifications to the request. No cloning occurs, and the request is not
+///      consumed.
+///
+/// Additionally, note that `LocalRequest` implements `Clone`. As such, if the
+/// same request needs to be dispatched multiple times, the request can first be
+/// cloned and then dispatched: `request.clone().dispatch()`.
+///
+/// [`Client`]: crate::local::blocking::Client
+/// [`header`]: #method.header
+/// [`add_header`]: #method.add_header
+/// [`cookie`]: #method.cookie
+/// [`remote`]: #method.remote
+/// [`body`]: #method.body
+/// [`set_body`]: #method.set_body
+/// [`dispatch`]: #method.dispatch
+/// [`mut_dispatch`]: #method.mut_dispatch
+pub struct LocalRequest<'c> {
+    inner: crate::local::LocalRequest<'c>,
+    client: &'c Client,
+}
+
+impl<'c> LocalRequest<'c> {
+    pub(crate) fn new(
+        client: &'c Client,
+        method: Method,
+        uri: Cow<'c, str>
+    ) -> LocalRequest<'c> {
+        let inner = crate::local::LocalRequest::new(&client.inner, method, uri);
+        Self { inner, client }
+    }
+
+    /// Retrieves the inner `Request` as seen by Rocket.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use rocket::local::blocking::Client;
+    ///
+    /// let client = Client::new(rocket::ignite()).expect("valid rocket");
+    /// let req = client.get("/");
+    /// let inner_req = req.inner();
+    /// ```
+    #[inline]
+    pub fn inner(&self) -> &Request<'c> {
+        self.inner.inner()
+    }
+
+    fn _map<F>(self, f: F) -> Self
```

Comment:
This seems like a big dance to do. See my top-level comment.

Reply:
I think this is just an artifact of an older, much uglier, workaround to make `block_on` re-entrant and I'll remove it.